### PR TITLE
Changes to the calculation of speed and flow aim difficulty

### PR DESCRIPTION
**Pull request by TextAdventurer12**
This contains a handful of changes:
- an attempt to stop the double-dipping of spaced streams into speed and aim pp by reducing aim pp if spacing is less than the single spacing bonus
- a nerf to the speed bonus
- an angle bonus for distance in speed calculation
- an increase to strain decay base for speed, resulting in a buff to longer streams

Values when combined with the other difficulty calculation changes (#28843 + #28844 + #28845 + #28846) can be seen [here](https://pp.huismetbenen.nl/rankings/players/kermit-rebalance) (onion access only)

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs [hunk 1]
```diff
@@ -54,12 +55,23 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
...
-            double distance = Math.Min(single_spacing_threshold, travelDistance + osuCurrObj.MinimumJumpDistance);
+            double distance = (travelDistance + osuCurrObj.MinimumJumpDistance) * angleBonus;
+            double distanceBonus = distance_multiplier * Math.Min(1, Math.Pow(distance / single_spacing_threshold, 3.5));
 
-            return (speedBonus + speedBonus * Math.Pow(distance / single_spacing_threshold, 3.5)) * doubletapness / strainTime;
+            double speedDifficulty = (speedBonus + distanceBonus) * doubletapness / strainTime;
+
+            return speedDifficulty;
+        }
+        private static double calculateAngleBonus(double angle)
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs:69]: The idea of increasing distance by angle means that acute angles in streams are considered to be flowed in a circular-ish way (which they usually aren't except rare potentially-snappable cases), therefore this angle bonus makes distance evaluation abusable (you can make a pattern that requires way smaller path than evaluated) and I think that's not the direction we should go.

**Review by stanriders — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs (hunk 1 above)
Comment (TextAdventurer12) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs]: I'm now using your angular velocity solution to this

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs [hunk 2]
```diff
@@ -118,6 +118,14 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
...
 
+            // The spacing bonus in speed evaluation
+            double flowBonus = Math.Pow((osuLastObj?.MinimumJumpDistance?? 0) / SpeedEvaluator.single_spacing_threshold, 3.5);
+            // Part of the aiming difficulty for this object is accounted for in the speed evaluator, so reduce aim difficulty here
+            if (flowBonus < 1)
+            {
+                aimStrain *= 0.5 + 0.5 * Math.Sqrt(flowBonus);
+                wideAngleBonus *= 0.5 + 0.5 * Math.Sqrt(flowBonus);
+            }
             // Add in acute angle bonus or wide angle bonus + velocity change bonus, whichever is larger.
```
Comment (tsunyoku) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:129]:
```suggestion
            }

            // Add in acute angle bonus or wide angle bonus + velocity change bonus, whichever is larger.
```

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs [hunk 3]
```diff
@@ -54,12 +55,25 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
...
-            double distance = Math.Min(single_spacing_threshold, travelDistance + osuCurrObj.MinimumJumpDistance);
+            double distance = travelDistance + osuCurrObj.MinimumJumpDistance;
+            double distanceBonus = distance_multiplier * Math.Min(1, Math.Pow(distance / single_spacing_threshold, 3.5));
 
-            return (speedBonus + speedBonus * Math.Pow(distance / single_spacing_threshold, 3.5)) * doubletapness / strainTime;
+            double adjustedDistanceScale = 1.0;
+
+            if (osuCurrObj.Angle.HasValue && osuPrevObj?.Angle != null && osuCurrObj.Angle != osuPrevObj.Angle)
+            {
+
```
Comment (tsunyoku) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs:68]:
```suggestion
```

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs [hunk 4]
```diff
@@ -54,12 +55,25 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
...
+                double angleDifference = Math.Abs(osuCurrObj.Angle.Value - osuPrevObj.Angle.Value);
+                double angleDifferenceAdjusted = Math.Sin((angleDifference) / 2) * 180.0;
+                double angularVelocity = angleDifferenceAdjusted / (0.1 * strainTime);
+                double angularVelocityBonus = Math.Max(0.0, Math.Pow(angularVelocity, 0.4) - 1.0); //Math.Max(0.0, 1.0 - 1.0 / angularVelocity);
```
Comment (tsunyoku) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/SpeedEvaluator.cs:72]:
```suggestion
                double angularVelocityBonus = Math.Max(0.0, Math.Pow(angularVelocity, 0.4) - 1.0);
```

**Review by tsunyoku — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by smoogipoo** [comment]
!diffcalc

**Comment by TextAdventurer12** [comment]
moved back to draft while I work through some feedback

**Comment by TextAdventurer12** [comment]
closing this because it underwent changes on another branch alongside some other changes

**Pull request closed without merging**